src/Pos/Txp/Toil/Balances/Functions.hs

```diff
@@ -0,0 +1,94 @@
+                                      getTxDistribution, txOutStake)
+import           Pos.Txp.Toil.Class  (MonadBalances (..), MonadBalancesRead (..))
+
+type BalancesMode m = (MonadBalances m, WithLogger m)
+
+-- | Apply transactions to balances.
+applyTxsToBalances :: BalancesMode m => [(TxAux, TxUndo)] -> m ()
+applyTxsToBalances txun = do
+    let (txOutPlus, txInMinus) = concatStakes txun
+    recomputeStakes txOutPlus txInMinus
+
+-- | Rollback application of transactions to balances.
+rollbackTxsBalances :: BalancesMode m => [(TxAux, TxUndo)] -> m ()
+rollbackTxsBalances txun = do
+    let (txOutMinus, txInPlus) = concatStakes txun
+    recomputeStakes txInPlus txOutMinus
+
+----------------------------------------------------------------------------
+-- Helpers
+----------------------------------------------------------------------------
+
+-- Compute new stakeholder's stakes by lists of spent and received coins.
+recomputeStakes
+    :: BalancesMode m
+    => [(StakeholderId, Coin)]
+    -> [(StakeholderId, Coin)]
+    -> m ()
+recomputeStakes plusDistr minusDistr = do
+    let needResolve =
+            HS.toList $
+            HS.fromList (map fst plusDistr) `HS.union`
+            HS.fromList (map fst minusDistr)
+    resolvedStakes <- mapM resolve needResolve
+    totalStake <- getTotalStake
+    let positiveDelta = sumCoins (map snd plusDistr)
+    let negativeDelta = sumCoins (map snd minusDistr)
+    let newTotalStake = unsafeIntegerToCoin $
+                        coinToInteger totalStake + positiveDelta - negativeDelta
+
+    let newStakes
+          = HM.toList $
+              calcNegStakes minusDistr
+                  (calcPosStakes $ zip needResolve resolvedStakes ++ plusDistr)
+    setTotalStake newTotalStake
+    mapM_ (uncurry setStake) newStakes
+  where
+    createInfo = sformat ("Stake for " %build%" will be created in UtxoDB")
+    resolve ad = maybe (mkCoin 0 <$ logInfo (createInfo ad)) pure =<< getStake ad
+    calcPosStakes distr = foldl' plusAt HM.empty distr
+    calcNegStakes distr hm = foldl' minusAt hm distr
+    -- @pva701 says it's not possible to get negative coin here. We *can* in
+    -- theory get overflow because we're adding and only then subtracting,
+    -- but in practice it won't happen unless someone has 2^63 coins or
+    -- something.
+    plusAt hm (key, val) = HM.insertWith unsafeAddCoin key val hm
+    minusAt hm (key, val) =
+        HM.alter (maybe err (\v -> Just (unsafeSubCoin v val))) key hm
+      where
+        err = error ("recomputeStakes: no stake for " <> show key)
+
+-- Concatenate stakes of the all passed transactions and undos.
+concatStakes
+    :: [(TxAux, TxUndo)]
+    -> ([(StakeholderId, Coin)], [(StakeholderId, Coin)])
+concatStakes (unzip -> (txas, undo)) = (txasTxOutDistr, undoTxInDistr)
+  where
+    txasTxOutDistr = concatMap concatDistr txas
+    undoTxInDistr = concatMap txOutStake (fold $ map toList undo)
```

Comment:
`fold $ map toList undo` is `foldMap toList undo`. Maybe using `HLint` could help and we need _CodecClimate_